src/shims/fs.rs

```diff
@@ -257,6 +264,197 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         this.try_unwrap_io_result(result)
     }
 
+    fn statx(
+        &mut self,
+        dirfd_op: OpTy<'tcx, Tag>, // Has type `int`
+        pathname_op: OpTy<'tcx, Tag>, // Has type `const char *`
+        flags_op: OpTy<'tcx, Tag>, // Has type `int`
+        _mask_op: OpTy<'tcx, Tag>, // Has type `unsigned int`
+        statxbuf_op: OpTy<'tcx, Tag> // Has type `struct statx *`
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        this.check_no_isolation("statx")?;
+
+        let statxbuf_scalar = this.read_scalar(statxbuf_op)?.not_undef()?;
+        let pathname_scalar = this.read_scalar(pathname_op)?.not_undef()?;
+
+        // If the statxbuf or pathname pointers are null, the function fails with `EFAULT`.
+        if this.is_null(statxbuf_scalar)? || this.is_null(pathname_scalar)? {
+            let efault = this.eval_libc("EFAULT")?;
+            this.set_last_error(efault)?;
+            return Ok(-1);
+        }
+
+        // If statxbuf is not null, we need to produce a proper `MemPlace` to write the results of
+        // the function to it. This is needed because all the `statx` parameters are provided as
+        // `isize`s instead of having the proper types. In particular `statxbuf` should be a `*mut
+        // statx`. If this wasn't the case, we could use `deref_operand` instead.
```

Comment:
Okay I think we are getting there, but the comment is still somewhat confusing. First, the "if it is not null part" is redundant, the code says that quite clearly.

But I think the elephant in the room is that the comment doesn't say what we'd do usually: we'd just `deref_operand(statxbuf_op)`. But we can't as it's not of pointer type. So we do this manually and supply the pointer type.

Reply:
I think I'm not understanding. Line 292 says that we could use `deref_operand`

Reply:
Ah, you're putting the most important information last. Sorry I missed that. Please open with this:
* We'd like to use `deref_operand`.
* But we can't, because X.
* Hence we do Y.

Reply:
Heh sure no problem